src/viam/sdk/common/utils.cpp

```diff
@@ -55,5 +55,21 @@ std::string bytes_to_string(std::vector<unsigned char> const& b) {
     return img_string;
 };
 
+std::chrono::duration<int64_t, std::micro> from_proto(const google::protobuf::Duration& proto) {
+    return std::chrono::microseconds(
+        (int64_t)std::ceil(proto.seconds() * 1e6 + proto.nanos() / 1e3));
```

Comment:
This seems like there is a lot going on. There is `std::ceil`, floating point multiplications and divisions, etc.

Can you simplify it by decomposing the `protos.seconds` and `protos.nanos` separately into the `chrono` types and then combining?

```
... from_proto(Duration& proto) {
    namespace sc  = std::chrono;

    const sc::seconds seconds_part{proto.seconds()};
    const sc::nanoseconds nanos_part{proto.nanos()};
    
    const sc::microseconds from_seconds = std::duration_cast<sc::microseconds>(seconds_part);

    sc::microseconds from_nanos = 
std::duration_cast<sc::microseconds>(nanos_part);

    if ((nanos_part < 0) && (from_nanos > nanos_part))
        from_nanos -= 1;
    else if ((nanos_part > 0) && (from_nanos < nanos_part))
        from_nanos += 1;

    return from_seconds + from_nanos;
}
```

Rounding above is untested but inspired by
- https://en.cppreference.com/w/cpp/chrono/duration/ceil
- https://en.cppreference.com/w/cpp/chrono/duration/floor

Unfortunately, we can't use those because they are C++17 only.

Also, I'd definitely like to see some explicit unit tests for these conversions.

   
    

}

Reply:
Good idea. This is significantly cleaner.